### ISSUE TITLE
Stress-test backend

### ISSUE THREAD

**Issue by dakom** [ji-devs/ji-cloud#2139]
When I ran a batch process to execute a lot of backend requests concurrently, I got a 500 error. The workaround was I had to get rid of or slow down the concurrency.

I'd strongly suggest writing a simple tool that stress-tests the backend and that it should be run before going live

Here's some reference code, fwiw: https://github.com/ji-devs/ji-cloud/tree/sandbox/build-utils/legacy-player/transcode/src/create_jig

(this was written as just a one-off tool, i.e. doesn't handle errors elegantly and I sometimes adapted it with comments instead of real config options, just a reference for passing a reqwest client down to a bunch of async fns and having a max pool active at a time.)

Note that @mehcode said:

> this [500] error means that the "max connections" param in the app doesn't match the max connections of your DB

**Comment by dakom** [comment]
@johnnynotsolucky suggested looking into https://github.com/fcsonline/drill which might avoid the need for custom tools

**Comment by johnnynotsolucky** [comment]
Notes RE max_connections:

- Currently the app is configured for [5 connections](https://github.com/ji-devs/ji-cloud/blob/sandbox/backend/core/src/config.rs#L18)
- The release database is configured for 100 connections:

```
SELECT current_setting('max_connections');
 current_setting
-----------------
 100
(1 row)
```
- Cloud SQL configures max_connections on the database automatically based on available memory: https://cloud.google.com/sql/docs/postgres/flags#postgres-m

**Comment by johnnynotsolucky** [comment]
I've bumped the psql instance on sandbox up a notch (2vcpus lol) so that it can handle a slightly higher load, and reliably more connections. The previous 1vcpu instance was wholly insufficient whereby it could only handle < 50 concurrent requests for an easy workload, such as only loading JIG modules:

```yaml
  - name: Load JIG modules
    request:
      url: /v1/jig/<id>/live/module/{{item}}
    with_items:
      - <ids>
```

Rudimentary load testing on the backend now yields these results for 150 concurrent requests for a total of 13000 requests over ~2 minutes:

```
Auth user                 Total requests            1000
Auth user                 Successful requests       1000
Auth user                 Failed requests           0
Auth user                 Median time per request   1400ms
Auth user                 Average time per request  1689ms
Auth user                 Sample standard deviation 1018ms

Get profile               Total requests            1000
Get profile               Successful requests       1000
Get profile               Failed requests           0
Get profile               Median time per request   1204ms
Get profile               Average time per request  1244ms
Get profile               Sample standard deviation 749ms

Load metadata             Total requests            1000
Load metadata             Successful requests       1000
Load metadata             Failed requests           0
Load metadata             Median time per request   4342ms
Load metadata             Average time per request  4137ms
Load metadata             Sample standard deviation 2054ms

Load a JIGs data          Total requests            1000
Load a JIGs data          Successful requests       1000
Load a JIGs data          Failed requests           0
Load a JIGs data          Median time per request   2114ms
Load a JIGs data          Average time per request  2235ms
Load a JIGs data          Sample standard deviation 583ms

Load JIG modules          Total requests            9000
Load JIG modules          Successful requests       9000
Load JIG modules          Failed requests           0
Load JIG modules          Median time per request   878ms
Load JIG modules          Average time per request  937ms
Load JIG modules          Sample standard deviation 454ms
``` 

CPU usage on the cloudsql instance was close to 100%

Increasing concurrency from 150 results in requests failing intermittently - either timing out on the client (10s+) or possibly running out of available psql connections (however, for the above test, Cloud Run used ~29 containers with each instance having max psql connections of 10, cloudsql reported ~177 active connections of a max of 400).

Testing the `/v1/jig` endpoint for searching JIGs shows that > 8 concurrent requests takes longer than 10s per request - The higher the concurrency, the longer the response time with just a single request running around ~2 to ~4s.

```yaml
  - name: Search for JIGs
    request:
      url: /v1/jig?page=0&jigFocus=modules
```

I can't really see why search is slow - clousql shows ~17% cpu usage for 16 concurrent requests, but no other resource metrics moved by much. cloud run only showed a spike in request latencies (~32s/request) but also no resource metrics spiked.

Worth noting that fetching a JIGs data with `/v1/jig/<id>/live` suffers a similar effect, but less drastic - with 128 concurrent requests, the response time increases to ~8s.

---

I used [drill](https://github.com/fcsonline/drill) which isn't as powerful as something like Locust, but is sufficient for now I think.

**Comment by dakom** [comment]
@johnnynotsolucky amazing detective work!!!!

Great point about scaling up the DB. It is one (only?) part of the stack that does not scale automatically and does need to be increased manually.

@corinnewo @MendyBerger  this is something that probably needs to be explained to Chana/Josh, that there should be a pre-approval to scale the DB in anticipation of heavy usage, and @johnnynotsolucky is 100% right that the current settings were only sufficient for internal minimal users, not any sort of real live traffic, and should be increased to whatever amounts the dev team recommendeds even before minimal public release.

@johnnynotsolucky - search is hitting Algolia too I suppose and it might be hard to pinpoint exactly why it's slow on their side..

**Comment by johnnynotsolucky** [comment]
@dakom RE algolia - I'm fairly confident that they can handle the request load with ease - searching small datasets of a few 100k records is very fast with algolia. The jigzi search endpoint calls algolia as a first step and then loads the JIG data from the DB using the JIG IDs returned from algolia. The search against algolia first likely only adds a small amount of latency, but nothing which would account for these response times.

**Comment by dakom** [comment]
Yeah, I agree, it's unlikely that Algolia is the bottleneck. Though loading a Jig directly from an ID also sounds very cheap 🤷

**Comment by johnnynotsolucky** [comment]
After #2463 with a similar load test as before:

```
Auth user                 Total requests            350
Auth user                 Successful requests       350
Auth user                 Failed requests           0
Auth user                 Median time per request   2385ms
Auth user                 Average time per request  2405ms
Auth user                 Sample standard deviation 1267ms

Get profile               Total requests            350
Get profile               Successful requests       350
Get profile               Failed requests           0
Get profile               Median time per request   423ms
Get profile               Average time per request  579ms
Get profile               Sample standard deviation 453ms

Load metadata             Total requests            350
Load metadata             Successful requests       350
Load metadata             Failed requests           0
Load metadata             Median time per request   910ms
Load metadata             Average time per request  1103ms
Load metadata             Sample standard deviation 841ms

Search for JIGs           Total requests            350
Search for JIGs           Successful requests       350
Search for JIGs           Failed requests           0
Search for JIGs           Median time per request   447ms
Search for JIGs           Average time per request  499ms
Search for JIGs           Sample standard deviation 197ms

Load a JIGs data          Total requests            350
Load a JIGs data          Successful requests       350
Load a JIGs data          Failed requests           0
Load a JIGs data          Median time per request   396ms
Load a JIGs data          Average time per request  504ms
Load a JIGs data          Sample standard deviation 269ms

Load JIG modules          Total requests            3150
Load JIG modules          Successful requests       3150
Load JIG modules          Failed requests           0
Load JIG modules          Median time per request   801ms
Load JIG modules          Average time per request  829ms
Load JIG modules          Sample standard deviation 401ms
```

**Comment by johnnynotsolucky** [comment]
Removing the `high` label for now, but this task should remain open for monitoring.

**Comment by johnnynotsolucky** [comment]
On sandbox, at least, calls to browse are down from a p95 of 1.34s to a p95 of 155ms:

### Before
![image](https://user-images.githubusercontent.com/4161106/157403922-afd3b748-9963-4743-962a-81f7f81ba054.png)

### After
![image](https://user-images.githubusercontent.com/4161106/157403989-0da8aa87-418b-48f6-99ca-987aafb7868c.png)
